File: backend/app/protocol.py
"""
IPC Protocol v1.0 - Python Implementation
Matches specification in IPC-001
"""

from pydantic import BaseModel, Field
from typing import Optional, Any, List, Tuple
from enum import Enum
import time
import uuid


class MessageType(str, Enum):
    CMD_HANDSHAKE = "CMD_HANDSHAKE"
    RES_HANDSHAKE = "RES_HANDSHAKE"
    CMD_PING = "CMD_PING"
    RES_PONG = "RES_PONG"
    CMD_EXTRACT_EVIDENCE = "CMD_EXTRACT_EVIDENCE"
    CMD_PARSE_TABLE = "CMD_PARSE_TABLE"
    RES_SUCCESS = "RES_SUCCESS"
    RES_PROGRESS = "RES_PROGRESS"
    RES_ERROR = "RES_ERROR"
    CMD_SHUTDOWN = "CMD_SHUTDOWN"
    RES_ACK = "RES_ACK"


class CacheHit(str, Enum):
    MEMORY = "memory"
    DISK = "disk"
    MISS = "miss"


class ErrorSeverity(str, Enum):
    FATAL = "fatal"
    ERROR = "error"
    WARNING = "warning"
    INFO = "info"


class IpcMessage(BaseModel):
    """Universal message envelope"""
    protocol_v: str = "1.0.0"
    msg_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    timestamp: int = Field(default_factory=lambda: int(time.time() * 1000))
    type: MessageType
    payload: Any

    class Config:
        use_enum_values = True


class BoundingBox(BaseModel):
    """PDF bounding box in specified units"""
    x: float
    y: float
    width: float
    height: float
    unit: str = "pt"


class HandshakeRequestPayload(BaseModel):
    rust_version: str
    expected_protocol_v: str
    capabilities_requested: List[str]


class HandshakeResponsePayload(BaseModel):
    worker_pid: int
    docling_version: str
    python_version: str
    capabilities_supported: List[str]
    max_memory_mb: int
    status: str


class ExtractEvidencePayload(BaseModel):
    file_path: str
    page_index: int
    bbox: BoundingBox
    dpi: int = 150
    output_format: Optional[str] = "jpeg"
    quality: Optional[int] = 85


class ParseTablePayload(BaseModel):
    file_path: str
    page_index: int
    hint_bbox: Optional[BoundingBox] = None
    detection_confidence_threshold: float = 0.7
    language: str = "vie"


class SuccessPayload(BaseModel):
    req_id: str
    data: Any
    metadata: Optional[Any] = None


class ErrorPayload(BaseModel):
    req_id: str
    code: str
    severity: ErrorSeverity
    message: str
    details: Optional[Any] = None
    suggested_action: Optional[str] = None
    timestamp: int = Field(default_factory=lambda: int(time.time() * 1000))
    stack_trace: Optional[str] = None


class ProgressPayload(BaseModel):
    req_id: str
    stage: str
    current: int
    total: int
    stage_description: Optional[str] = None
    eta_seconds: Optional[int] = None


class ErrorCodes:
    """Standard error codes for IPC"""
    E_SYS_001 = "E-SYS-001"  # Internal System Error
    E_FILE_001 = "E-FILE-001"  # File Not Found
    E_VAL_001 = "E-VAL-001"  # Validation Error
    E_PROC_001 = "E-PROC-001"  # Processing Error


# Helper functions
def create_message(msg_type: MessageType, payload: Any) -> IpcMessage:
    """Create a new IPC message with auto-generated ID and timestamp"""
    return IpcMessage(type=msg_type, payload=payload)


def create_error_message(
    request_id: str,
    error_code: str,
    message: str,
    severity: ErrorSeverity = ErrorSeverity.ERROR,
    details: Optional[Any] = None,
    suggested_action: Optional[str] = None,
) -> IpcMessage:
    """Create an error response message"""
    payload = ErrorPayload(
        req_id=request_id,
        code=error_code,
        severity=severity,
        message=message,
        details=details,
        suggested_action=suggested_action,
    )
    # Using model_dump (pydantic v2) or dict (pydantic v1) for payload
    payload_data = payload.model_dump() if hasattr(payload, "model_dump") else payload.dict()
    return create_message(MessageType.RES_ERROR, payload_data)


def create_success_message(request_id: str, data: Any, metadata: Optional[Any] = None) -> IpcMessage:
    """Create a success response message"""
    payload = SuccessPayload(req_id=request_id, data=data, metadata=metadata)
    payload_data = payload.model_dump() if hasattr(payload, "model_dump") else payload.dict()
    return create_message(MessageType.RES_SUCCESS, payload_data)


# Add method to IpcMessage for easier serialization
def to_json(self) -> str:
    return self.model_dump_json() if hasattr(self, "model_dump_json") else self.json()

IpcMessage.to_json = to_json

File: crates/tachfileto-core/src/ipc/protocol.rs
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum MessageType {
    CmdHandshake,
    ResHandshake,
    CmdPing,
    ResPong,
    CmdExtractEvidence,
    CmdParseTable,
    ResSuccess,
    ResProgress,
    ResError,
    CmdShutdown,
    ResAck,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum CacheHit {
    Memory,
    Disk,
    Miss,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IpcMessage<T> {
    pub protocol_v: String,
    pub msg_id: Uuid,
    pub timestamp: i64,
    #[serde(rename = "type")]
    pub msg_type: MessageType,
    pub payload: T,
}

// Payload for CMD_HANDSHAKE
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HandshakeRequestPayload {
    pub rust_version: String,
    pub expected_protocol_v: String,
    pub capabilities_requested: Vec<String>,
}

// Payload for RES_HANDSHAKE
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HandshakeResponsePayload {
    pub worker_pid: u32,
    pub docling_version: String,
    pub python_version: String,
    pub capabilities_supported: Vec<String>,
    pub max_memory_mb: u32,
    pub status: String,
}

// Payload for CMD_EXTRACT_EVIDENCE
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExtractEvidencePayload {
    pub file_path: PathBuf,
    pub page_index: usize,
    pub bbox: BoundingBox,
    pub dpi: u16,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_format: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub quality: Option<u8>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BoundingBox {
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
    #[serde(default = "default_unit")]
    pub unit: String,
}

fn default_unit() -> String {
    "pt".to_string()
}

// Payload for CMD_PARSE_TABLE
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParseTablePayload {
    pub file_path: PathBuf,
    pub page_index: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hint_bbox: Option<BoundingBox>,
    #[serde(default = "default_confidence")]
    pub detection_confidence_threshold: f32,
    #[serde(default = "default_language")]
    pub language: String,
}

fn default_confidence() -> f32 {
    0.7
}

fn default_language() -> String {
    "vie".to_string()
}

// Payload for RES_SUCCESS
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SuccessPayload<T> {
    pub req_id: Uuid,
    pub data: T,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Value>,
}

// Payload for RES_ERROR
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorPayload {
    pub req_id: Uuid,
    pub code: String,
    pub severity: ErrorSeverity,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub suggested_action: Option<String>,
    pub timestamp: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stack_trace: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ErrorSeverity {
    Fatal,
    Error,
    Warning,
    Info,
}

// Payload for RES_PROGRESS
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProgressPayload {
    pub req_id: Uuid,
    pub stage: String,
    pub current: u32,
    pub total: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stage_description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub eta_seconds: Option<u32>,
}

// Helper to create messages quickly
impl<T: Serialize> IpcMessage<T> {
    pub fn new(msg_type: MessageType, payload: T) -> Self {
        Self {
            protocol_v: "1.0.0".to_string(),
            msg_id: Uuid::new_v4(),
            timestamp: chrono::Utc::now().timestamp_millis(),
            msg_type,
            payload,
        }
    }

    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }
}

impl<'de, T: Deserialize<'de>> IpcMessage<T> {
    pub fn from_json(json_str: &'de str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json_str)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_message_serialization() {
        let payload = HandshakeRequestPayload {
            rust_version: "1.83.0".to_string(),
            expected_protocol_v: "1.0.0".to_string(),
            capabilities_requested: vec!["ocr".to_string()],
        };

        let msg = IpcMessage::new(MessageType::CmdHandshake, payload);
        let json = serde_json::to_string(&msg).unwrap();

        assert!(json.contains("CMD_HANDSHAKE"));
        assert!(json.contains("protocol_v"));
        assert!(json.contains("1.0.0"));
    }
}
